File: scripts/gpu_step.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $GPUREG = "perl ./gpureg.pl";

sub g {
  my ($cmd) = @_;
  system("$GPUREG $cmd") == 0 or die "Failed: $cmd\n";
}

sub do_steps {
  my ($n) = @_;
  for my $i (1..$n) {
    g("step");
  }
  g("dbg");
}

print "=== GPU Interactive Stepper ===\n";
print "Commands:\n";
print "  <n>       step N cycles then show PC/instr\n";
print "  d <addr>  read DMEM[addr]\n";
print "  done      check done flag\n";
print "  dbg       show PC + IF_INSTR\n";
print "  q         quit\n\n";

while (1) {
  print "step> ";
  my $line = <STDIN>;
  last unless defined $line;
  chomp $line;
  $line =~ s/^\s+|\s+$//g;

  next if $line eq '';

  if ($line eq 'q' || $line eq 'quit') {
    last;

  } elsif ($line =~ /^(\d+)$/) {
    my $n = int($1);
    if ($n < 1) { print "Enter a positive integer.\n"; next; }
    print "--- stepping $n cycle(s) ---\n";
    do_steps($n);

  } elsif ($line =~ /^d\s+(\S+)$/) {
    g("dmem_read $1");

  } elsif ($line eq 'done') {
    g("done_check");

  } elsif ($line eq 'dbg') {
    g("dbg");

  } else {
    print "Unknown command: $line\n";
  }
}

print "Bye.\n";

File: scripts/python/gpu_step.py
#!/usr/bin/env python
import os
import subprocess
import sys

THIS_DIR = os.path.dirname(os.path.abspath(__file__))
GPUREG = os.path.join(THIS_DIR, "gpureg.py")
PYTHON = sys.executable or "python"

def write_line(text):
    sys.stdout.write("%s\n" % text)


def run_process(argv):
    return subprocess.Popen(argv).wait()


try:
    input_func = raw_input
except NameError:
    input_func = input


def g(*args):
    argv = [PYTHON, GPUREG] + list(args)
    if run_process(argv) != 0:
        raise SystemExit("Failed: %s" % " ".join(argv))


def do_steps(count):
    for _ in range(count):
        g("step")
    g("dbg")


def main():
    write_line("=== GPU Interactive Stepper ===")
    write_line("Commands:")
    write_line("  <n>       step N cycles then show PC/instr")
    write_line("  d <addr>  read DMEM[addr]")
    write_line("  done      check done flag")
    write_line("  dbg       show PC + IF_INSTR")
    write_line("  q         quit")
    write_line("")

    while True:
        try:
            line = input_func("step> ")
        except EOFError:
            break

        line = line.strip()
        if not line:
            continue

        if line in ("q", "quit"):
            break
        elif line.isdigit():
            count = int(line)
            if count < 1:
                write_line("Enter a positive integer.")
                continue
            write_line("--- stepping %s cycle(s) ---" % count)
            do_steps(count)
        elif line.startswith("d "):
            parts = line.split(None, 1)
            if len(parts) == 2:
                g("dmem_read", parts[1])
            else:
                write_line("Unknown command: %s" % line)
        elif line == "done":
            g("done_check")
        elif line == "dbg":
            g("dbg")
        else:
            write_line("Unknown command: %s" % line)

    write_line("Bye.")


if __name__ == "__main__":
    main()
